ucp/application_context.py

```diff
@@ -0,0 +1,376 @@
+        msg,
+        len(msg),
+        ep._ctrl_tag_recv,
+        pending_msg=ep.pending_msg_list[-1],
+    )
+
+    shutdown_fut.add_done_callback(partial(handle_ctrl_msg, weakref.ref(ep), log, msg))
+
+
+def listener_handler(ucp_endpoint, ctx, worker, func, guarantee_msg_order):
+    async def run(ucp_endpoint, ctx, worker, func, guarantee_msg_order):
+        loop = asyncio.get_event_loop()
+        # TODO: exceptions in this callback is never showed when no
+        #       get_exception_handler() is set.
+        #       Is this the correct way to handle exceptions in asyncio?
+        #       Do we need to set this in other places?
+        if loop.get_exception_handler() is None:
+            loop.set_exception_handler(asyncio_handle_exception)
+
+        # We create the Endpoint in three steps:
+        #  1) Generate unique IDs to use as tags
+        #  2) Exchange endpoint info such as tags
+        #  3) Use the info to create a new Endpoint
+        msg_tag = hash(uuid.uuid4())
+        ctrl_tag = hash(uuid.uuid4())
+        peer_info = await exchange_peer_info(
+            ucp_endpoint=ucp_endpoint,
+            msg_tag=msg_tag,
+            ctrl_tag=ctrl_tag,
+            guarantee_msg_order=guarantee_msg_order,
+        )
+        ep = public_api.Endpoint(
+            ucp_endpoint=ucp_endpoint,
+            worker=worker,
+            ctx=ctx,
+            msg_tag_send=peer_info["msg_tag"],
+            msg_tag_recv=msg_tag,
+            ctrl_tag_send=peer_info["ctrl_tag"],
+            ctrl_tag_recv=ctrl_tag,
+            guarantee_msg_order=guarantee_msg_order,
+        )
+
+        logging.debug(
+            "listener_handler() server: %s, msg-tag-send: %s, "
+            "msg-tag-recv: %s, ctrl-tag-send: %s, ctrl-tag-recv: %s"
+            % (
+                hex(ucp_endpoint.handle),
+                hex(ep._msg_tag_send),
+                hex(ep._msg_tag_recv),
+                hex(ep._ctrl_tag_send),
+                hex(ep._ctrl_tag_recv),
+            )
+        )
+
+        # Setup the control receive
+        setup_ctrl_recv(ep)
+
+        # Removing references here to avoid delayed clean up
+        del ctx
+
+        # Finally, we call `func` asynchronously (even if it isn't coroutine)
+        if asyncio.iscoroutinefunction(func):
+            await func(ep)
+        else:
+
+            async def _func(ep):  # coroutine wrapper
+                func(ep)
+
+            await _func(ep)
+
+    asyncio.ensure_future(run(ucp_endpoint, ctx, worker, func, guarantee_msg_order))
+
+
+async def _non_blocking_mode(weakref_ctx):
+    """This help function maintains a UCX progress loop.
+    Notice, it only keeps a weak reference to `ApplicationContext`, which makes it
+    possible to call `ucp.reset()` even when this loop is running.
+    """
+    while True:
+        ctx = weakref_ctx()
+        if ctx is None:
+            return
+        ctx.progress()
+        del ctx
+        await asyncio.sleep(0)
```

Comment:
Couple questions. Why do we need the explicit `del`? Also why is `sleep(0)` needed?

Reply:
We need to delete `ctx` to avoid holding a reference to `ApplicationContext`, which would make `ucp.reset()` fails.

We use `sleep(0)` to give other co-routines a chance to run.

Reply:
Does `0` give them enough of a chance? Should this be something larger?